src/storage/mvcc/consistency_check.rs

```diff
@@ -0,0 +1,462 @@
+const SAFE_POINT_WINDOW: usize = 120;
+
+// When leader broadcasts a ComputeHash command to followers, it's possible that the safe point
+// becomes stale when the command reaches followers. So use a 2 minutes window to reduce this.
+fn get_safe_point_for_check(mut safe_point: u64) -> u64 {
+    safe_point >>= PHYSICAL_SHIFT_BITS;
+    safe_point += (SAFE_POINT_WINDOW * 1000) as u64; // 120s * 1000ms/s.
+    safe_point << PHYSICAL_SHIFT_BITS
+}
+
+#[derive(Clone)]
+pub struct Mvcc<E: KvEngine> {
+    _engine: PhantomData<E>,
+    local_safe_point: Arc<AtomicU64>,
+}
+
+impl<E: KvEngine> Coprocessor for Mvcc<E> {}
+
+impl<E: KvEngine> Mvcc<E> {
+    pub fn new(safe_point: Arc<AtomicU64>) -> Self {
+        Mvcc {
+            _engine: Default::default(),
+            local_safe_point: safe_point,
+        }
+    }
+}
+
+impl<E: KvEngine> ConsistencyCheckObserver<E> for Mvcc<E> {
+    fn update_context(&self, context: &mut Vec<u8>) -> bool {
+        context.push(ConsistencyCheckMethod::Mvcc as u8);
+        context.reserve(8);
+        let len = context.len();
+
+        let mut safe_point = self.local_safe_point.load(AtomicOrdering::Acquire);
+        safe_point = get_safe_point_for_check(safe_point);
+        unsafe {
+            context.set_len(len + 8);
+            std::ptr::copy_nonoverlapping(
+                safe_point.to_le_bytes().as_ptr(),
+                &mut context[len] as _,
+                8,
+            );
+        }
+        // Skiped all other observers.
+        true
+    }
+
+    fn compute_hash(
+        &self,
+        region: &kvproto::metapb::Region,
+        context: &mut &[u8],
+        snap: &E::Snapshot,
+    ) -> Result<Option<u32>> {
+        if context.is_empty() {
+            return Ok(None);
+        }
+        assert_eq!(context[0], ConsistencyCheckMethod::Mvcc as u8);
+        let safe_point = u64::from_le_bytes(context[1..9].try_into().unwrap());
+        *context = &context[9..];
+
+        let local_safe_point = self.local_safe_point.load(AtomicOrdering::Acquire);
+        if safe_point < local_safe_point {
+            warn!(
+                "skip consistency check"; "region_id" => region.get_id(),
+                "safe_ponit" => safe_point,
+                "local_safe_point" => local_safe_point,
+            );
+            return Ok(None);
+        }
+
+        let mut scanner = MvccInfoScanner::new(
+            |cf, opts| snap.iterator_cf_opt(cf, opts).map_err(|e| box_err!(e)),
+            Some(&keys::data_key(region.get_start_key())),
+            Some(&keys::data_end_key(region.get_end_key())),
+            MvccChecksum::new(safe_point),
+        )?;
+        while scanner.next_item()?.is_some() {}
+
+        // Computes the hash from the Region state too.
+        let mut digest = scanner.observer.digest;
+        let region_state_key = keys::region_state_key(region.get_id());
+        digest.update(&region_state_key);
+        match snap.get_value_cf(CF_RAFT, &region_state_key) {
```

Comment:
Why not scan_cf from snapshot and update digest

Reply:
It's what `ConsistencyCheckMethod::Raw` does. Considering `delete_files_in_range` and `compaction filter`, the raw checksums between leader and followers can be different.